crates/isograph_compiler/src/write_artifacts.rs

```diff
@@ -7,63 +8,134 @@ use std::{
+
+        for relative_path in to_add.into_iter() {
+            if let Some((absolute_path, content)) = artifact_map.remove(&relative_path) {
+                artifacts_to_disk
+                    .artifacts_to_write
+                    .insert(absolute_path, content);
+            }
+        }
+
+        artifacts_to_disk.artifacts_to_delete = to_delete
+            .into_iter()
+            .map(|path| artifact_directory.join(path))
+            .collect();
+        artifacts_to_disk.cleanup_artifact_directory = false;
+    }
+
+    *file_system_state = new_file_system_state;
+
+    artifacts_to_disk
+}
+
+#[tracing::instrument(skip(artifacts_to_disk, artifact_directory))]
+pub(crate) fn write_artifacts_to_disk(
+    artifacts_to_disk: ChangedArtifacts,
+    artifact_directory: &PathBuf,
 ) -> DiagnosticResult<usize> {
-    if artifact_directory.exists() {
+    if artifact_directory.exists() && artifacts_to_disk.cleanup_artifact_directory {
         fs::remove_dir_all(artifact_directory).map_err(|e| {
             unable_to_do_something_at_path_diagnostic(
                 artifact_directory,
                 &e.to_string(),
                 "delete directory",
             )
         })?;
+
+        fs::create_dir_all(artifact_directory).map_err(|e| {
+            let message = e.to_string();
+            unable_to_do_something_at_path_diagnostic(
+                artifact_directory,
+                &message,
+                "create directory",
+            )
+        })?;
     }
-    fs::create_dir_all(artifact_directory).map_err(|e| {
-        let message = e.to_string();
-        unable_to_do_something_at_path_diagnostic(artifact_directory, &message, "create directory")
-    })?;
 
     let mut count = 0;
-    for path_and_content in paths_and_contents {
-        // Is this better than materializing paths_and_contents sooner?
+    for (path, content) in artifacts_to_disk.artifacts_to_write.iter() {
         count += 1;
 
-        let absolute_directory = match path_and_content.artifact_path.type_and_field {
-            Some(type_and_field) => artifact_directory
-                .join(type_and_field.parent_object_entity_name.lookup())
-                .join(type_and_field.selectable_name.lookup()),
-            None => artifact_directory.clone(),
-        };
+        let absolute_directory = path.parent().expect("path must have a parent");
+
         fs::create_dir_all(&absolute_directory).map_err(|e| {
             unable_to_do_something_at_path_diagnostic(
-                &absolute_directory,
+                &absolute_directory.to_path_buf(),
                 &e.to_string(),
                 "create directory",
             )
         })?;
 
-        let absolute_file_path =
-            absolute_directory.join(path_and_content.artifact_path.file_name.lookup());
-        let mut file = File::create(&absolute_file_path).map_err(|e| {
+        let mut file = File::create(&path).map_err(|e| {
```

Comment:
Another thing I was planning to do (or maybe we can do it as a follow up), is to write to a tmp file first and then do atomic replace. Some say it is a proper way to update files that are heavily used by another service, e.g. lsp